libs/ledger/src/consensus/consensus.cpp

```diff
@@ -0,0 +1,203 @@
+                             current_block_number_ + aeon_period_ + 1);
+  }
+}
+
+NextBlockPtr Consensus::GenerateNextBlock()
+{
+  NextBlockPtr ret;
+
+  // Number of block we want to generate
+  uint64_t const block_number = current_block_number_ + 1;
+  uint64_t       entropy      = 0;
+
+  // Try to get entropy for the block we are generating - is allowed to fail if we request too
+  // early
+  if (EntropyGeneratorInterface::Status::OK !=
+      beacon_->GenerateEntropy(current_block_.body.hash, block_number, entropy))
+  {
+    return ret;
+  }
+
+  // Note here the previous block's entropy determines miner selection
+  if (!stake_->ShouldGenerateBlock(current_block_, mining_address_))
+  {
+    return ret;
+  }
+
+  ret = std::make_unique<Block>();
+
+  ret->body.previous_hash = current_block_.body.hash;
+  ret->body.block_number  = block_number;
+  ret->body.miner         = mining_address_;
+  ret->body.entropy       = entropy;
+  ret->weight             = stake_->GetBlockGenerationWeight(current_block_, mining_address_);
+
+  return ret;
+}
+
+Status Consensus::ValidBlock(Block const &previous, Block const &current)
+{
+  Status ret = Status::YES;
+
+  // Attempt to ascertain the beacon value within the block
+  uint64_t entropy = 0;
+
+  // Try to get entropy for the block we are generating - because of races it could be
+  // we have yet to receive it from the network.
+  auto result = beacon_->GenerateEntropy(current.body.hash, current.body.block_number, entropy);
+
+  switch (result)
+  {
+  case EntropyGeneratorInterface::Status::NOT_READY:
+    FETCH_LOG_INFO(LOGGING_NAME, "Too early for entropy for block: ", current.body.block_number);
+    ret = Status::UNKNOWN;
+    break;
+  case EntropyGeneratorInterface::Status::OK:
+    if (entropy == current.body.entropy)
+    {
+      ret = Status::YES;
+    }
+    else
+    {
+      ret = Status::NO;
+    }
+    break;
+  case EntropyGeneratorInterface::Status::FAILED:
+    ret = Status::NO;
+    break;
+  }
+
+  if (ret == Status::NO)
+  {
+    FETCH_LOG_WARN(LOGGING_NAME, "Saw incorrect random beacon from: ",
+                   current_block_.body.miner.address().ToBase64(),
+                   ".Block number: ", current_block_.body.block_number, " expected: ", entropy,
+                   " got: ", current_block_.body.entropy);
+    return Status::NO;
+  }
+
+  // Here we assume the entropy is correct and proceed to verify against the stake subsystem
+  if (!stake_->ValidMinerForBlock(previous, current.body.miner))
+  {
+    FETCH_LOG_WARN(LOGGING_NAME, "Received block has incorrect miner identity set");
+    return Status::NO;
+  }
+
+  if (current.weight != stake_->GetBlockGenerationWeight(previous, current.body.miner) &&
+      current.weight != 0)
+  {
+    FETCH_LOG_WARN(LOGGING_NAME,
+                   "Incorrect stake weight found for block. Weight: ", current.weight);
+    return Status::NO;
+  }
+
+  return Status::YES;
+}
+
+void Consensus::Refresh()
```

Comment:
?

Reply:
I was going to have this basically look at the main chain or other structures to see if it's got anything interesting (the suggestion that DKG has failed and you should abort). However I cut this PR short since I wanted to merge in your muddle changes and avoid conflicts